File: 2015/R/d10.R
d <- strsplit(readLines("../inputs/input_10.txt"), "")[[1]]

f <- function(d, n) {
  for (i in seq_len(n - 1)) {
    d <- rle(d)
    d <- c(rbind(d$lengths, d$values))
  }
  d <- rle(d)
  length(d$lengths) + length(d$values)
}

c(f(d, 40), f(d, 50))

File: 2015/python/d10.py
def read(filename):
    for line in open(filename, 'r'):
        return line.replace('\n', '')

def proc(s):
    res = ''
    prev = s[0]
    count = 1
    for i in range(1, len(s)):
        if (s[i] != prev):
            res += str(count) + prev
            prev = s[i]
            count = 1
        else:
            count += 1
    return(res + str(count) + prev)

def tests():
    assert(proc('1') == '11')
    assert(proc('11') == '21')
    assert(proc('21') == '1211')
    assert(proc('1211') == '111221')
    assert(proc('111221') == '312211')

def loop(d, n):
    for i in range(0, n):
        d = proc(d)
    return len(d)

def part1(d):
    return loop(d, 40)

def part2(d):
    return loop(d, 50)

tests()
input = read('../inputs/input_10.txt')
print(part1(input))
print(part2(input))
